libevmasm/SemanticInformation.cpp

```diff
@@ -278,12 +281,41 @@ bool SemanticInformation::invalidatesStorage(Instruction _instruction)
 	case Instruction::CREATE:
 	case Instruction::CREATE2:
 	case Instruction::SSTORE:
-		return true;
+		return yul::SideEffects::Write;
+
+	case Instruction::SLOAD:
+		return yul::SideEffects::Read;
+
 	default:
-		return false;
+		return yul::SideEffects::Pure;
 	}
 }
 
+yul::SideEffects::Effect SemanticInformation::otherState(Instruction _instruction)
+{
+	switch (_instruction)
+	{
+	case Instruction::CALL:
+	case Instruction::CALLCODE:
+	case Instruction::DELEGATECALL:
+	case Instruction::CREATE:
+	case Instruction::CREATE2:
+	case Instruction::SSTORE:
+		return yul::SideEffects::Write;
+
+	case Instruction::EXTCODESIZE:
+	case Instruction::EXTCODEHASH:
+	case Instruction::RETURNDATASIZE:
+	case Instruction::BALANCE:
+	case Instruction::SELFBALANCE:
```

Comment:
Add:
 - returndatacopy
 - extcodecopy
 - staticcall

Add a comment:
 - pc and gas are specifically excluded here
 - things like caller, callvalue, address are excluded because they cannot change during execution

Reply:
@chriseth gas and pc cannot be excluded. This is because if they end up having `None` as `Effect`, then they will get moved. Example: https://github.com/ethereum/solidity/pull/9312/files#r450033224

I think we need to have list of 'exceptions'. This should include pc, gas, invalid and stop.

Reply:
Yes, there is still 'movable' and 'sideEffectFree'. Instructions like `return` and `stop` should both be non-movable and not side-effect-free.